src/rules/indentRule.ts

```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    if (root.getEnd() === pos) {
+        return root;
+    }
+    let result: ts.Node | undefined;
+    ts.forEachChild(root, function cb(node: ts.Node): void {
+        if (node.getFullStart() < pos && node.getEnd() >= pos) {
+            result = recursiveGetNodeAt(node, pos);
+        }
+        ts.forEachChild(node, cb);
+    });
+    return result;
+}
+
+function getNodeDeepth(node: ts.Node, line: string): number {
+    let result = 0;
+    let parent: ts.Node | undefined = node.parent;
+    const blockTypes: ts.SyntaxKind[] = [
+        ts.SyntaxKind.Block,
+        ts.SyntaxKind.Parameter,
+        ts.SyntaxKind.CaseBlock,
+        ts.SyntaxKind.EnumMember,
+        ts.SyntaxKind.CaseClause,
+        ts.SyntaxKind.JsxElement,
+        ts.SyntaxKind.ModuleBlock,
+        ts.SyntaxKind.TemplateSpan,
+        ts.SyntaxKind.DefaultClause,
+        ts.SyntaxKind.ClassDeclaration,
+        ts.SyntaxKind.ArrayLiteralExpression,
+        ts.SyntaxKind.ObjectLiteralExpression,
+    ];
+    if (isCloseElement(node, line)) {
+        result --;
+    }
+    if (isCloseParen(node, line)) {
+        result ++;
+    }
+    if (inStatementParen(node)) {
+        result ++;
+    }
+    while (parent !== undefined) {
+        if (blockTypes.indexOf(parent.kind) > -1) {
+            result++;
+        }
+        if (inStatementParen(parent)) {
+            result ++;
+        }
+        parent = parent.parent;
+    }
+    if (nodeAtOutside(node)) {
+        result--;
+    }
+    if (
+        node.parent !== undefined
+        && node.parent.kind === ts.SyntaxKind.CallExpression
+        && /^\s*\./.test(line)
+    ) {
+        result ++;
+    }
+    return result;
+}
+
+function inStatementParen(node: ts.Node): boolean {
+    if (node.parent === undefined) {
+        return false;
+    }
+    const parentInstatment = [
+        ts.SyntaxKind.IfStatement,
+        ts.SyntaxKind.WhileStatement,
+        ts.SyntaxKind.ForStatement,
+        ts.SyntaxKind.ForOfStatement,
+        ts.SyntaxKind.DoStatement,
+        ts.SyntaxKind.WithStatement,
+    ].indexOf(node.parent.kind) >= 0;
+    return node.kind !== ts.SyntaxKind.Block && parentInstatment;
+}
+
+function isCloseParen(node: ts.Node, line: string): boolean {
+    if (node.kind !== ts.SyntaxKind.Block) {
+        return false;
+    }
+    return /^\s*\)/.test(line);
+}
+
+function isCloseElement(node: ts.Node, line: string): boolean {
+    if (node.parent === undefined) {
+        return false;
+    }
+    return node.parent.kind === ts.SyntaxKind.JsxElement
+        && /^\s*<\//.test(line);
+}
+
+function inStringTemplate(node: ts.Node | undefined): boolean {
+    if (node === undefined) {
+        return false;
+    }
+    return node.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral
+    || (
+        node.kind >= ts.SyntaxKind.FirstTemplateToken
+        && node.kind <= ts.SyntaxKind.LastTemplateToken
+    );
+}
+
+function nodeAtOutside(node: ts.Node): boolean {
```

Comment:
please add a comment to describe what this function does

Reply:
done